### PR TITLE
Sprint 4 Development

### PR DESCRIPTION
**Pull request by brianfunk**
# Sprint 4 Development including:
## v0.0.9
#85 Implement searching contact as part of POST API
#139 Create files route to get files from S3 for outfitters
#152 Update outfitter route to “temp-outfitters”
#153 Allow region, district, and forest codes to start with 0
#158 Create contact as part of POST request
#168 Remove fields from tempOutfitters
#169 Update naming in validationSchema for clarity
#170 Update scemaValidation initiation to be iterative
#171 Update error reporting to use standard error function
#172 Ensure all errors are returned
#174 Retrieve all Non-SUDS application fields stored in database
#176 Upload validation - file size
#178 Updates to POST validation based on updated mapping file
#180 Create a mock basic API service
#181 Create POST requests for contact address
#182 Make routing schema-based
#183 Create POST request for contact phone
#185 Save Non-SUDS fields in middle-layer database
#186 Save files to middle-layer database
#187 Upload and validate files to S3
#192 Update postSchema fields
#193 Update environment variables for database
#196 Add file validation properties to validation schema
#206 Update package dba to run sequelize locally*
#210 Document process for creating new permit type
#211 Automate JSDoc output

### CODE REVIEW

File: seeders/applications.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+require('dotenv').config();
+
+module.exports = {
+	up: function (queryInterface, Sequelize) { 
+		return queryInterface.bulkInsert('applications', [
+			{
+				id: Sequelize.literal('uuid_generate_v4()'), control_number: '123456789', form_name: 'FS-2700-3f', region: '38', forest: '50', district: '50', website: 'www.sample.com',
```
Comment (lauraGgit) [inline seeders/applications.js]: It's not actually necessary to seed these applications right? its just test data?

File: seeders/users.js [hunk 2]
```diff
@@ -5,13 +5,13 @@ require('dotenv').config();
 module.exports = {
 	up: function (queryInterface, Sequelize) { 
 		return queryInterface.bulkInsert('users', [
-		        {user_name: process.env.ADMINROLE_USER, pass_hash: process.env.ADMINROLE_HASH, user_role: "admin", created: "now()", updated: "now()" }, 
-		        {user_name: process.env.USERROLE_USER, pass_hash: process.env.USERROLE_HASH, user_role: "user", created: "now()", updated: "now()" } 
+			{user_name: process.env.ADMINROLE_USER, pass_hash: process.env.ADMINROLE_HASH, user_role: 'admin', created: 'now()', updated: 'now()'},
```
Comment (lauraGgit) [inline seeders/users.js]: I might have missed it, but were you able to add the create user command?

File: controllers/basic.js [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+			fulfill(res.contact);
+		})
+		.catch(function(err){
+			reject(err);
+		});
+	});
+}
+
+/** Sends requests needed to create an application via the Basic API
+ * @param  {Object} req - Request Object
```
Comment (cameronwolf) [inline controllers/basic.js]: Yes that is the incoming request from the intake module.

**Comment by lauraGgit** [comment]
Also outline in `README` Basic Url and how to set it as the mocks.

File: migrations/01-create-filetypes.js [hunk 4]
```diff
@@ -5,11 +5,14 @@ module.exports = {
 			file_type: {
```
Comment (lpendem) [inline migrations/01-create-filetypes.js]: @lauraGgit sure, we will plan to do sequelize alter process when this changes next time.

File: controllers/validation.js [hunk 5]
```diff
@@ -0,0 +1,626 @@
+/*
+
+  ___ ___       ___               _ _       _   ___ ___ 
+ | __/ __|  ___| _ \___ _ _ _ __ (_) |_    /_\ | _ \_ _|
+ | _|\__ \ / -_)  _/ -_) '_| '  \| |  _|  / _ \|  _/| | 
+ |_| |___/ \___|_| \___|_| |_|_|_|_|\__| /_/ \_\_| |___|
+
+*/
+
+//*******************************************************************
```
Comment (janwilli) [inline controllers/validation.js]: This file includes both validation and error handling. Splitting it up might result in confusion.

File: seeders/users.js (hunk 2 above)
Comment (janwilli) [inline seeders/users.js]: No, but we do have subtask #207 in the backlog to address this.

File: controllers/basic.js [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+//*******************************************************************
+// required modules
+const request = require('request-promise');
+
+//*******************************************************************
+// other files
+
+const db = require('./db.js');
+const error = require('./error.js');
+const basicURL = process.env.BASICURL;
```
Comment (janwilli) [inline controllers/basic.js]: We added [#253](https://github.com/nci-ats/fs-middlelayer-api/issues/253) to address this.

File: controllers/basic.js [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+					obj[requestToUse][key] = field[key];
+					requests.push(obj);
+				}
+			}
+		});
+	});
+	requests.forEach((request)=>{
+		const key = Object.keys(request)[0];
+		const obj = {};
+		obj[key] = {};
```
Comment (janwilli) [inline controllers/basic.js]: We added [#254](https://github.com/nci-ats/fs-middlelayer-api/issues/254) to address this.

File: controllers/basic.js [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+		jsonResponse.success = true;
+		jsonResponse.api = 'FS ePermit API';
+		jsonResponse.type = 'controller';
+		jsonResponse.verb = req.method;
+		jsonResponse.src = 'json';
+		jsonResponse.route = req.originalUrl;
+		jsonResponse.controlNumber = controlNumber;
+		jsonResponse.apiRequest = body;
+		jsonResponse.basicPosts = postObject;
+		return res.json(jsonResponse);
```
Comment (janwilli) [inline controllers/basic.js]: This will be addressed by #255.

File: controllers/db.js [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+const getApplication = function(controlNumber, callback){
+
+	models.applications.findOne({
+		where: {
+			control_number: controlNumber
+		}
+	}).then(function(appl) {
+		if (appl){
+			if (appl.form_name === 'FS-2700-3f') {
```
Comment (janwilli) [inline controllers/db.js]: We added [#252](https://github.com/nci-ats/fs-middlelayer-api/issues/252) to address this.

File: controllers/index.js [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+		}
+	}
+
+}
+
+/** If body passed in as string, converts it to a JSON object
+ * @param  {Object} req - request object
+ * @return {Object} - request body as a JSON Object
+ */
+function getBody(req){
```
Comment (janwilli) [inline controllers/index.js]: We added [#256](https://github.com/nci-ats/fs-middlelayer-api/issues/256) to address this.

File: controllers/index.js [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+		});
+	}
+	const errorMessage = validation.generateErrorMesage(allErrors);
+	if (allErrors.errorArray.length !== 0){
+		return error.sendError(req, res, 400, errorMessage, allErrors.errorArray);
+	}
+	else {
+		const toStoreInDB = db.getDataToStoreInDB(sch, body);
+
+		const controlNumber = (Math.floor((Math.random() * 10000000000) + 1)).toString(); //TODO: remove - used for mocks
```
Comment (janwilli) [inline controllers/index.js]: We added [#257](https://github.com/nci-ats/fs-middlelayer-api/issues/257) to address this.

File: controllers/index.js [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+						basic.postToBasic(req, res, sch, body, controlNumber);
+						
+					}
+				});
+			}
+		});
+	}
+};
+
+const use = function(req, res){
```
Comment (janwilli) [inline controllers/index.js]: We added [#258](https://github.com/nci-ats/fs-middlelayer-api/issues/258) to address this.

File: controllers/validation.js [hunk 13]
```diff
@@ -0,0 +1,626 @@
...
+			case 'properties':
+				findField(schema.properties, fieldCopy, func);
+				break;
+			}
+		}
+	});
+}
+
+function handleMissingError(output, result, counter, schema){
+	requiredFields = [];
```
Comment (janwilli) [inline controllers/validation.js]: We added [#259](https://github.com/nci-ats/fs-middlelayer-api/issues/259) to address this.

**Comment by lauraGgit** [comment]
Closed in favor of the purely Sprint #4 Commits in #13.

**Pull request closed without merging**